File: native/comparison/fastedit.c
#define _POSIX_C_SOURCE 200809L

#include <errno.h>
#include <fcntl.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

enum { READ_CHUNK_SIZE = 1 << 20 };

static void fail(const char *message) {
	perror(message);
	exit(errno ? errno : 1);
}

static void fail_message(const char *message) {
	fprintf(stderr, "%s\n", message);
	exit(1);
}

static char *read_all(const char *path, size_t *size_out) {
	int fd = open(path, O_RDONLY | O_CLOEXEC);
	if (fd < 0) fail("open input");

	struct stat st;
	if (fstat(fd, &st) != 0) fail("fstat input");

	size_t capacity = st.st_size > 0 ? (size_t)st.st_size + 1 : READ_CHUNK_SIZE;
	char *buffer = malloc(capacity);
	if (!buffer) fail("malloc input");

	size_t size = 0;
	for (;;) {
		if (size == capacity) {
			size_t next_capacity = capacity < READ_CHUNK_SIZE ? READ_CHUNK_SIZE : capacity * 2;
			char *next = realloc(buffer, next_capacity);
			if (!next) fail("realloc input");
			buffer = next;
			capacity = next_capacity;
		}

		ssize_t bytes_read = read(fd, buffer + size, capacity - size);
		if (bytes_read == 0) break;
		if (bytes_read < 0) {
			if (errno == EINTR) continue;
			fail("read input");
		}
		size += (size_t)bytes_read;
	}

	close(fd);
	buffer[size] = '\0';
	*size_out = size;
	return buffer;
}

static void write_all_fd(int fd, const char *buffer, size_t size) {
	size_t written = 0;
	while (written < size) {
		ssize_t bytes_written = write(fd, buffer + written, size - written);
		if (bytes_written < 0) {
			if (errno == EINTR) continue;
			fail("write output");
		}
		written += (size_t)bytes_written;
	}
}

int main(int argc, char **argv) {
	if (argc != 4) {
		fprintf(stderr, "usage: %s <target> <old-text-file> <new-text-file>\n", argv[0]);
		return 1;
	}

	const char *target_path = argv[1];
	const char *old_text_path = argv[2];
	const char *new_text_path = argv[3];

	size_t target_size = 0;
	size_t old_size = 0;
	size_t new_size = 0;
	char *target = read_all(target_path, &target_size);
	char *old_text = read_all(old_text_path, &old_size);
	char *new_text = read_all(new_text_path, &new_size);

	if (old_size == 0) fail_message("oldText must not be empty");

	char *first = NULL;
	for (size_t i = 0; i + old_size <= target_size; i += 1) {
		if (memcmp(target + i, old_text, old_size) == 0) {
			if (first) fail_message("oldText not unique");
			first = target + i;
		}
	}

	if (!first) fail_message("oldText not found");

	size_t prefix_size = (size_t)(first - target);
	size_t suffix_offset = prefix_size + old_size;
	size_t suffix_size = target_size - suffix_offset;
	size_t output_size = prefix_size + new_size + suffix_size;

	char tmp_path[4096];
	int tmp_len = snprintf(tmp_path, sizeof(tmp_path), "%s.tmp.%ld", target_path, (long)getpid());
	if (tmp_len <= 0 || (size_t)tmp_len >= sizeof(tmp_path)) fail_message("temporary path too long");

	int out_fd = open(tmp_path, O_WRONLY | O_CREAT | O_TRUNC | O_CLOEXEC, 0644);
	if (out_fd < 0) fail("open tmp output");

	write_all_fd(out_fd, target, prefix_size);
	write_all_fd(out_fd, new_text, new_size);
	write_all_fd(out_fd, target + suffix_offset, suffix_size);

	if (fsync(out_fd) != 0) fail("fsync output");
	if (close(out_fd) != 0) fail("close output");
	if (rename(tmp_path, target_path) != 0) fail("rename output");

	free(target);
	free(old_text);
	free(new_text);

	printf("{\"ok\":true,\"bytes\":%zu}\n", output_size);
	return 0;
}

File: native/fastedit.zig
const std = @import("std");
const Io = std.Io;
const posix = std.posix;
const system = posix.system;

fn usage(argv0: []const u8) noreturn {
    std.debug.print("usage: {s} <target> <old-text-file> <new-text-file>\n", .{argv0});
    std.process.exit(1);
}

fn fatal(comptime message: []const u8, err: anyerror) noreturn {
    std.debug.print(message ++ ": {s}\n", .{@errorName(err)});
    std.process.exit(1);
}

fn fatalMessage(comptime message: []const u8) noreturn {
    std.debug.print(message ++ "\n", .{});
    std.process.exit(1);
}

fn fatalErrno(comptime message: []const u8, err: std.posix.E) noreturn {
    std.debug.print(message ++ ": {s}\n", .{@tagName(err)});
    std.process.exit(1);
}

fn writeAllFd(fd: i32, content: []const u8) void {
    var written: usize = 0;
    while (written < content.len) {
        const rc = system.write(fd, content[written..].ptr, content.len - written);
        switch (posix.errno(rc)) {
            .SUCCESS => {
                const n: usize = @intCast(rc);
                if (n == 0) fatalErrno("write output", .IO);
                written += n;
            },
            .INTR => {},
            else => |err| fatalErrno("write output", err),
        }
    }
}

fn toPosixPath(path: []const u8) [std.posix.PATH_MAX - 1:0]u8 {
    return posix.toPosixPath(path) catch fatal("path", error.NameTooLong);
}

fn unlinkPath(path: []const u8) void {
    const path_z = toPosixPath(path);
    _ = system.unlink(&path_z);
}

fn renamePath(old_path: []const u8, new_path: []const u8) void {
    const old_z = toPosixPath(old_path);
    const new_z = toPosixPath(new_path);
    const rc = system.renameat(posix.AT.FDCWD, &old_z, posix.AT.FDCWD, &new_z);
    switch (posix.errno(rc)) {
        .SUCCESS => {},
        else => |err| {
            unlinkPath(old_path);
            fatalErrno("rename output", err);
        },
    }
}

fn writeJson(bytes: usize) void {
    var buf: [96]u8 = undefined;
    const text = std.fmt.bufPrint(&buf, "{{\"ok\":true,\"bytes\":{}}}\n", .{bytes}) catch unreachable;
    writeAllFd(1, text);
}

fn fsyncParentDir(path: []const u8) void {
    const slash = std.mem.lastIndexOfScalar(u8, path, '/');
    const dir = if (slash) |idx| (if (idx == 0) path[0..1] else path[0..idx]) else ".";
    const fd = posix.openat(posix.AT.FDCWD, dir, .{ .ACCMODE = .RDONLY, .DIRECTORY = true, .CLOEXEC = true }, 0) catch return;
    _ = system.fsync(fd);
    _ = system.close(fd);
}

fn writeContent(fd: posix.fd_t, target: []const u8, new_text: []const u8, prefix_size: usize, suffix_offset: usize) void {
    writeAllFd(fd, target[0..prefix_size]);
    writeAllFd(fd, new_text);
    writeAllFd(fd, target[suffix_offset..]);

    switch (posix.errno(system.fsync(fd))) {
        .SUCCESS => {},
        else => |err| fatalErrno("fsync output", err),
    }
    switch (posix.errno(system.close(fd))) {
        .SUCCESS => {},
        else => |err| fatalErrno("close output", err),
    }
}

fn writeOutput(target_path: []const u8, tmp_path: []const u8, target: []const u8, new_text: []const u8, prefix_size: usize, suffix_offset: usize, mode: posix.mode_t) void {
    const fd = posix.openat(posix.AT.FDCWD, tmp_path, .{ .ACCMODE = .WRONLY, .CREAT = true, .TRUNC = true, .CLOEXEC = true }, mode) catch |err| fatal("open tmp output", err);
    switch (posix.errno(system.fchmod(fd, mode))) {
        .SUCCESS => {},
        else => |err| fatalErrno("fchmod output", err),
    }
    writeContent(fd, target, new_text, prefix_size, suffix_offset);
    renamePath(tmp_path, target_path);
    fsyncParentDir(target_path);
}

fn writeThrough(target_path: []const u8, target: []const u8, new_text: []const u8, prefix_size: usize, suffix_offset: usize) void {
    const fd = posix.openat(posix.AT.FDCWD, target_path, .{ .ACCMODE = .WRONLY, .TRUNC = true, .CLOEXEC = true }, 0) catch |err| fatal("open target", err);
    writeContent(fd, target, new_text, prefix_size, suffix_offset);
}

pub fn main(init: std.process.Init) !void {
    const io = init.io;
    const arena = init.arena.allocator();
    const args = try init.minimal.args.toSlice(arena);
    if (args.len != 4) usage(args[0]);

    const target_path = args[1];
    const old_text_path = args[2];
    const new_text_path = args[3];

    const target = Io.Dir.cwd().readFileAlloc(io, target_path, arena, .limited(std.math.maxInt(usize))) catch |err| fatal("open input", err);
    const old_text = Io.Dir.cwd().readFileAlloc(io, old_text_path, arena, .limited(std.math.maxInt(usize))) catch |err| fatal("open input", err);
    const new_text = Io.Dir.cwd().readFileAlloc(io, new_text_path, arena, .limited(std.math.maxInt(usize))) catch |err| fatal("open input", err);

    if (old_text.len == 0) fatalMessage("oldText must not be empty");

    const first_index = std.mem.indexOf(u8, target, old_text) orelse fatalMessage("oldText not found");
    const second_start = first_index + old_text.len;
    if (std.mem.indexOf(u8, target[second_start..], old_text) != null) fatalMessage("oldText not unique");

    const suffix_offset = first_index + old_text.len;
    const output_size = first_index + new_text.len + (target.len - suffix_offset);

    const target_z = toPosixPath(target_path);
    var stx: system.Statx = undefined;
    switch (posix.errno(system.statx(posix.AT.FDCWD, &target_z, posix.AT.SYMLINK_NOFOLLOW, .{ .TYPE = true, .MODE = true }, &stx))) {
        .SUCCESS => {},
        else => |err| fatalErrno("stat target", err),
    }
    if (posix.S.ISLNK(stx.mode)) {
        writeThrough(target_path, target, new_text, first_index, suffix_offset);
        writeJson(output_size);
        return;
    }

    const tmp_path = std.fmt.allocPrint(arena, "{s}.tmp.{}", .{ target_path, system.getpid() }) catch |err| fatal("malloc", err);
    if (tmp_path.len > 4095) fatalMessage("temporary path too long");

    writeOutput(target_path, tmp_path, target, new_text, first_index, suffix_offset, @intCast(stx.mode & 0o777));
    writeJson(output_size);
}
